### DISCUSSION TITLE
NotImplementedError

### DISCUSSION THREAD

**Discussion by onlyonewater** [pyg-team/pytorch_geometric#8390] [Q&A]
Hi, pyg team, I am a freshman for pgy package, when I run the code, I get the error like this:
`
  File "/mnt/workspace/duhao/.conda/envs/diffdock_pp/lib/python3.10/site-packages/torch/nn/parallel/parallel_apply.py", line 61, in _worker
    output = module(*input, **kwargs)
  File "/mnt/workspace/duhao/.conda/envs/diffdock_pp/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/mnt/workspace/duhao/.conda/envs/diffdock_pp/lib/python3.10/site-packages/torch_geometric/nn/data_parallel.py", line 69, in forward
    inputs = self.scatter(data_list, self.device_ids)
  File "/mnt/workspace/duhao/.conda/envs/diffdock_pp/lib/python3.10/site-packages/torch_geometric/nn/data_parallel.py", line 77, in scatter
    count = torch.tensor([data.num_nodes for data in data_list])
  File "/mnt/workspace/duhao/.conda/envs/diffdock_pp/lib/python3.10/site-packages/torch_geometric/data/feature_store.py", line 515, in __iter__
    raise NotImplementedError
`
I do not why, since I use the data_parallel and the batch size is 4 with 4 GPUs

**Comment by rusty1s** [comment]
It looks like you are using `FeatureStore` in combination with `nn.DataParallel`. This is currently a mis-use, as `nn.DataParallel` cannot deal with single graphs (it expects a dataset of graphs instead). We have collected a few examples on how to use `NeighborLoader` on multi-GPU [here](https://github.com/pyg-team/pytorch_geometric/tree/master/examples/multi_gpu).

**Reply by onlyonewater** [reply]
but, it uses the datalistloader to convert multiple graphs to a large single graph, like the examples shown in: https://github.com/pyg-team/pytorch_geometric/blob/master/examples/multi_gpu/data_parallel.py

**Reply by rusty1s** [reply]
Can you show me an example script of yours?

**Reply by onlyonewater** [reply]
it seems that I want to use four GPUs in DataParallel with `model =DataParallel(model, device_ids=[4,5,6,7])`, but in the beginning code, I use `torch.cuda.set_device(4)`, so there exists a mismatch between the true GPU num (1) with my expected GPU numbers (4)? I think this may be the reason? I am not sure

**Reply by onlyonewater** [reply]
well, the code in https://github.com/ketatam/DiffDock-PP may be a little complicated.

**Reply by onlyonewater** [reply]
and now I confirm that using single GPU with DataParallel is OK, I use dataloader class in single GPU situation

**Reply by rusty1s** [reply]
Sorry for late reply. Is there any chance you can provide a minimal example here to reproduce? How does the input of the model look like? Note also that we deprecated `DataParallel` a while ago, so I strongly suggest to switch to `DistributedDataParallel` instead.
